### DISCUSSION TITLE
TaskGroup.start() like anyio has.

### DISCUSSION THREAD

**Discussion by jonathanslenders** [fastapi/asyncer#119] [Questions]
Follow-up of: https://twitter.com/tiangolo/status/1750533650201154032

Anyio has a `TaskGroup.start()` function alongside `TaskGroup.start_soon()`: https://anyio.readthedocs.io/en/stable/api.html#anyio.abc.TaskGroup.start. This allows for starting a task in a task group and wait for it to signal readiness.

This is really an important feature in structured concurrency. Because, sometimes certain operations, even when they are concurrent, are only supposed to start when other concurrent tasks are past a certain point of readiness. E.g., one task might start a web server, and the main task will wait for the web server to actually have opened a port for listening before proceeding.

I noticed that asyncer has a `soonify()` which corresponds to `start_soon`, but it does not have a `startify` or anything that would correspond to `start_soon`. You might think that instead passing an event to the task for it to signal readiness would be sufficient, but it's not. With an event, we would end up with the following:

```python
async def some_task(ready_event: Event) -> None:
    ...
    ready_event.set()
    ...

async with create_task_group() as tg:
    ready_event = Event()
    tg.start_soon(some_task, ready_event)
    await ready_event.wait()
```

That's already three lines of boilerplace to make `TaskGroup.start()` working, but it doesn't cover everything. If something happens and the task returns (without exception) before the event was set, then this code will hang forever. 

The big problem from a typing point of view is that `start()` takes a `TaskStatus`, as you can see here: https://anyio.readthedocs.io/en/stable/tasks.html#starting-and-initializing-tasks
That means, one argument is passed in by the `start()` helper itself, and the other arguments are passed in by the caller of the function.

Look at the anyio source code. `TaskStatus.start()` is in fact already fully typed since anyio 4.0, thanks to `typing.Unpack`:
https://github.com/agronholm/anyio/blob/master/src/anyio/abc/_tasks.py#L49
So, asyncer doesn't add much beyond support for keyword arguments.
`TaskStatus.start()` on the other hand is not very much typed, see: https://github.com/agronholm/anyio/blob/master/src/anyio/abc/_tasks.py#L66

So, to summarize, there are actually two points here:
- `TaskGroup.start` is important to have, it simplifies a lot of code.
- Typing `TaskGroup.start` seems non-trivial. I can't think of a way to reshape the call and make the typing work with paramspec/unpack/etc... as asyncer does with `soonify()`

**Comment by jonathanslenders** [comment]
I just realized we can do better in anyio too: https://github.com/agronholm/anyio/issues/678 It's totally doable to offer a strictly typed `TaskGroup.start()` implementation.

**Comment by YuriiMotov** [comment]
For those who have the same question.

There is still no `TaskGroup.startify()` method, but `asyncer.TaskGroup` is inherited from `anyio`'s `TaskGroup` and it has `start()` method which you can use:

```py
import anyio
from anyio.abc import TaskStatus
import asyncer


async def handler(stream):
    ...


async def start_some_service(
    port: int, *, task_status: TaskStatus[None] = anyio.TASK_STATUS_IGNORED
):
    async with await anyio.create_tcp_listener(
        local_host="127.0.0.1", local_port=port
    ) as listener:
        task_status.started()
        await listener.serve(handler)



async def get_data():
    async with asyncer.create_task_group() as task_group:
        await task_group.start(start_some_service, 5000)
        async with await anyio.connect_tcp("127.0.0.1", 5000) as stream:
            ...


async def main():
    await get_data()


anyio.run(main)
```

https://anyio.readthedocs.io/en/stable/tasks.html#starting-and-initializing-tasks
